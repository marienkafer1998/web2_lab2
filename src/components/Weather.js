import React from "react";
import WeatherParam from "./WeatherParam";
import Loader from "./Loader";

const API_ICON_URL = "https://openweathermap.org/img/wn/";

export function getIconURL(iconCode) {
    return `${API_ICON_URL}${iconCode}.png`;
}

export default class Weather extends React.Component {
    componentDidMount() {
        this.props.onFetch();
    }

    render() {
        console.log('In weather');
        if (!this.props.weather) {
            return this.renderLoader();
        }

        return this.renderWeather();
    }

    renderLoader() {
        return (
            <Loader/>
        )
    }

    renderWeather() {
        console.log(this.props.weather);
        const {
            weather: {
                cityName,
                temperature,
                icon,
                windSpeed,
                description,
                pressure,
                humidity,
                coords: {
                    lat: latitude,
                    lon: longitude
                } = {}
            }, onDelete
        } = this.props;
        return (
            <div className="weather">
                <div className="header">
                    <div className="city-name">{cityName}</div>
                    <div className="temperature">{temperature} &#8451;</div>
                    <img src={getIconURL(icon)} alt="Weather icon" />
                </div>

                <WeatherParam name="Wind" value={`${windSpeed} m/s`} />
                <WeatherParam name="Cloudness" value={description} />
                <WeatherParam name="Pressure" value={`${pressure} hPa`} />
                <WeatherParam name="Humidity" value={`${humidity}%`} />
                <WeatherParam name="Coords" value={`${latitude}, ${longitude}`} />
                {onDelete && <button onClick={onDelete}>X</button>}

            </div>
        );
    }

}

