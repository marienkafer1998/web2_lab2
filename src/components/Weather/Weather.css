.small-weather {
    font-size: 10px;
}

.main-weather {
    font-size: 20px;
}

.cityName {
    color: #74878c;
    font-weight: 500;
    font-size: 3em;
}

.image {
    width: 6em;
}

.temp {
    color: #81848a;
    font-weight: 500;
    font-size: 4em;
}

.card-img-top {
    flex: 0 0 auto;
}

.col-sm-4 {
    border: 1px solid
}