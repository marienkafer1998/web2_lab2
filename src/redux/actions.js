
const API_BASE_URL = "http://api.openweathermap.org/data/2.5/weather?APPID=c98ceead6e7b88c9c865eaf7bdbb291d"


export const Actions = {
    SET_GEOLOCATION: "SET_GEOLOCATION",
    GET_RESPONSE: "GET_RESPONSE",
    GET_RESPONSE_NAME: "GET_RESPONSE_NAME",
    ADD_FAVORITE: "ADD_FAVORITE",
    DELETE_FAVORITE: "DELETE_FAVORITE",
    }

export function setGeolocation(coords) {
    return {
        type: Actions.SET_GEOLOCATION,
        payload: coords
    }
}

export function loadWeatherWithCoords(coords) {
    const API_URL = `${API_BASE_URL}&lat=${coords.lat}&lon=${coords.lon}`;

    return function (dispatch) {
        fetch(API_URL)
            .then(response => {
                    response.json()
                        .then(json => {
                            console.log(json);
                            if (response.ok) {
                                dispatch(getResponseCoords(json));
                            }
                        })
                }
            )
    }
}


export function loadWeatherWithName(cityName) {
    const API_URL = `${API_BASE_URL}&q=${cityName}`;
    console.log(API_URL);

    return function(dispatch) {
        fetch(API_URL)
            .then(response => {
                    response.json()
                        .then(json => {
                            console.log(response, json);

                            if (json.cod === 200) {
                                console.log("City input ", cityName);

                                dispatch(getResponseName(cityName, json));
                            } else {
                                dispatch(deleteFavorite(cityName));
                            }
                        })
                }
            )
    }
}


export function getResponseCoords(response) {
    return {
        type: Actions.GET_RESPONSE,
        payload: response
    }
}

export function getResponseName(cityName, response ) {
    return {
        type: Actions.GET_RESPONSE_NAME,
        payload: {
            cityName,
            response
        }
    }
}


export function addFavorite(cityName) {
    return {
        type: Actions.ADD_FAVORITE,
        payload: cityName
    };
}

export function deleteFavorite(cityName) {
    return {
        type: Actions.DELETE_FAVORITE,
        payload: cityName
    };
}