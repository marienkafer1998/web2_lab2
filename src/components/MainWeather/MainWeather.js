import React from "react";
import WeatherParameter from "../WeatherParameter/WeatherParameter";
import Loader from "../Loader/Loader";
import "../Weather/Weather.css";
import {Col, Container, Row} from "reactstrap";

const API_ICON_URL = "https://openweathermap.org/img/wn/";

export function getIconURL(iconCode) {
    return `${API_ICON_URL}${iconCode}.png`;
}

export default class MainWeather extends React.Component {
    componentDidMount() {
        console.log('parasha', typeof this.props.onFetch);
        this.props.onFetch();
    }

    render() {
        if (!this.props.weather) {
            return <Loader/>

        }

        return this.renderWeather();
    }


    renderWeather() {

        const {
            weather: {
                cityName,
                temperature,
                icon,
                windSpeed,
                description,
                pressure,
                humidity,
                coords: {
                    lat: latitude,
                    lon: longitude
                } = {}
            },
        } = this.props;
        let  body;
        if (this.props.weather) {
            body = <Container className={"main-weather"}>
                <Row>
                    <Col>
                        <Row className={"cityName"}>{cityName}</Row>
                        <Row><Col>
                            <img className={"image"} src={getIconURL(icon)} alt="Weather icon" />
                        </Col>
                            <Col className="temp">
                                {(temperature - 273.15).toFixed(1).toString()}â„ƒ
                            </Col>
                        </Row>
                    </Col>
                    <Col>
                        <WeatherParameter name="Wind" value={`${windSpeed} m/s`} />
                        <WeatherParameter name="Cloudness" value={description} />
                        <WeatherParameter name="Pressure" value={`${pressure} hPa`} />
                        <WeatherParameter name="Humidity" value={`${humidity}%`} />
                        <WeatherParameter name="Coords" value={`${latitude}, ${longitude}`} />
                    </Col>

                </Row>
            </Container>;}
        return (
            <Container className="weather">
                {body}
            </Container>

        );
    }

}

