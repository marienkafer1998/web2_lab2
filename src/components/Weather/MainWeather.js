import React from "react";
import WeatherParam from "../WeatherParam/WeatherParam";
import Loader from "../Loader";
import "./Weather.css";
import {Button, Col, Container, Row} from "reactstrap";

const API_ICON_URL = "https://openweathermap.org/img/wn/";

export function getIconURL(iconCode) {
    return `${API_ICON_URL}${iconCode}.png`;
}

export default class MainWeather extends React.Component {
    componentDidMount() {
        this.props.onFetch();
    }

    render() {
        if (!this.props.weather) {
            return <Loader/>

        }

        return this.renderWeather();
    }


    renderWeather() {

        const {
            weather: {
                cityName,
                temperature,
                icon,
                windSpeed,
                description,
                pressure,
                humidity,
                coords: {
                    lat: latitude,
                    lon: longitude
                } = {}
            },
        } = this.props;
        let  body;
        if (this.props.weather) {
            body = <Container className={"main-weather"}>
                <Row>
                    <Col>
                        <Row className={"cityName"}>{cityName}</Row>
                        <Row><Col>
                            <img className={"image"} src={getIconURL(icon)} alt="Weather icon" />
                        </Col>
                            <Col className="temp">
                                {(temperature - 273.15).toFixed(1).toString()}â„ƒ
                            </Col>
                        </Row>
                    </Col>
                    <Col>
                        <WeatherParam name="Wind" value={`${windSpeed} m/s`} />
                        <WeatherParam name="Cloudness" value={description} />
                        <WeatherParam name="Pressure" value={`${pressure} hPa`} />
                        <WeatherParam name="Humidity" value={`${humidity}%`} />
                        <WeatherParam name="Coords" value={`${latitude}, ${longitude}`} />
                    </Col>

                </Row>
            </Container>;}
        return (
            <Container className="weather">
                {body}
            </Container>

        );
    }

}

