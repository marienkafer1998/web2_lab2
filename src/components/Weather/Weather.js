import React from "react";
import WeatherParameter from "../WeatherParameter/WeatherParameter";
import Loader from "../Loader/Loader";
import "./Weather.css";
import {Button, Col, Container, Row} from "reactstrap";

const API_ICON_URL = "https://openweathermap.org/img/wn/";

export function getIconURL(iconCode) {
    return `${API_ICON_URL}${iconCode}.png`;
}

export default class Weather extends React.Component {
    componentDidMount() {
        this.props.onFetch();
    }

    render() {
        if (!this.props.weather) {
            return <Loader/>

        } else {
            return this.renderWeather();
        }
    }


    renderWeather() {

        const {
            weather: {
                cityName,
                temperature,
                icon,
                windSpeed,
                description,
                pressure,
                humidity,
                coords: {
                    lat: latitude,
                    lon: longitude
                } = {}
            }, onDelete
        } = this.props;
        let header, body;
        if (this.props.weather) {
            header = <Container className={"small-weather"}>
                <Row>

                    <Col className={"cityName"}>
                        {cityName}
                    </Col>
                    <Col className={"temp"}>
                        {(temperature - 273.15).toFixed(1).toString()}â„ƒ
                    </Col>
                    <Col className={"image"}>
                        <img src={getIconURL(icon)} alt="Weather icon" />
                    </Col>
                    <Col>
                        {onDelete && <Button size={"10px"} className="button" onClick={onDelete}>X</Button>}
                    </Col>
                </Row>
            </Container>;
            body = <div>
                <WeatherParameter name="Wind" value={`${windSpeed} m/s`} />
                <WeatherParameter name="Cloudness" value={description} />
                <WeatherParameter name="Pressure" value={`${pressure} hPa`} />
                <WeatherParameter name="Humidity" value={`${humidity}%`} />
                <WeatherParameter name="Coords" value={`${latitude}, ${longitude}`} />
            </div>}
        return (
            <Container className="weather">
                {header}
                {body}
            </Container>

        );
    }

}

