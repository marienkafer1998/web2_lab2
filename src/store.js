import { createStore, combineReducers, applyMiddleware } from "redux";
import logger from "redux-logger";
import thunk from "redux-thunk";
import { LOCAL_STORAGE_KEY } from "./LocalStorage";

import geoReducer from "./redux/geoReducer";
import favReducer from "./redux/favReducer";

const store = createStore(
    combineReducers({geo: geoReducer, fav: favReducer}),
    applyMiddleware(logger, thunk)
);



store.subscribe(() => {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify([...store.getState().fav.favorites.keys()]));
});

export default store;
