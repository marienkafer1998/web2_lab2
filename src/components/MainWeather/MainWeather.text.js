import React from 'react';
import renderer from 'react-test-renderer';
import MainWeather from "./MainWeather.js";
import {Provider} from "react-redux";
import store from "../../store";
import {loadWeatherWithCoords} from "../../redux/actions";

const weatherData = {
    cityName: 'Moscow',
    temperature: "275.24",
    pressure: "1003",
    humidity: "69",
    windSpeed: "2",
    icon: "04n",
    description: "overcast clouds",
    coords: {lon: "55.75" , lat: "37.62"}
};


it('MainWeather functional rendered properly', () => {
    const tree = renderer.create(
        <Provider store={store}>
            <MainWeather onFetch={() => loadWeatherWithCoords(weatherData.coords)}
                         weather={weatherData}/>
        </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

