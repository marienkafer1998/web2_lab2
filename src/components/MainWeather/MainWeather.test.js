import React from 'react';
import renderer from 'react-test-renderer';
import MainWeather from "./MainWeather.js";
import {Provider} from "react-redux";
import store from "../../store";
import {loadWeatherWithCoords} from "../../redux/actions";
import Weather from "../Weather/Weather";

describe('MainWeather functional rendered properly', () => {

    const weatherData = {
        cityName: 'Moscow',
        temperature: "275.24",
        pressure: "1003",
        humidity: "69",
        windSpeed: "2",
        icon: "04n",
        description: "overcast clouds",
        coords: {lon: "55.75", lat: "37.62"}
    };


    test('should has city weather', () => {
        const tree = renderer.create(
            <Provider store={store}>
                <MainWeather onFetch={() => loadWeatherWithCoords(weatherData.coords)}
                             weather={weatherData}/>
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should has loader', () => {
        const tree = renderer
            .create(
                <Weather
                    onFetch={() => {
                    }}
                />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});