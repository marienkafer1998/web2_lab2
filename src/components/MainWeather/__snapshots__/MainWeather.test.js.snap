// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainWeather functional rendered properly should has city weather 1`] = `
<div
  className="weather container"
>
  <div
    className="main-weather container"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="cityName row"
        >
          Moscow
        </div>
        <div
          className="row"
        >
          <div
            className="col"
          >
            <img
              alt="Weather icon"
              className="image"
              src="https://openweathermap.org/img/wn/04n.png"
            />
          </div>
          <div
            className="temp col"
          >
            2.1
            â„ƒ
          </div>
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="param"
        >
          <div
            className="name"
          >
            Wind
          </div>
          <div
            className="value"
          >
            2 m/s
          </div>
        </div>
        <div
          className="param"
        >
          <div
            className="name"
          >
            Cloudness
          </div>
          <div
            className="value"
          >
            overcast clouds
          </div>
        </div>
        <div
          className="param"
        >
          <div
            className="name"
          >
            Pressure
          </div>
          <div
            className="value"
          >
            1003 hPa
          </div>
        </div>
        <div
          className="param"
        >
          <div
            className="name"
          >
            Humidity
          </div>
          <div
            className="value"
          >
            69%
          </div>
        </div>
        <div
          className="param"
        >
          <div
            className="name"
          >
            Coords
          </div>
          <div
            className="value"
          >
            37.62, 55.75
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MainWeather functional rendered properly should has loader 1`] = `
<div
  className="container"
>
  <h1>
    Please, wait, weather is loading...
  </h1>
  <img
    alt="Weather is loading"
    src="https://anatomised.com/wp-content/uploads/2016/05/spinner-test4.gif"
    width="100px"
  />
</div>
`;
